pkg/cli/gettoken/oidc/client.go

```diff
@@ -0,0 +1,229 @@
+package oidc
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	gooidc "github.com/coreos/go-oidc/v3/oidc"
+	"github.com/int128/oauth2cli"
+	"golang.org/x/oauth2"
+
+	"github.com/openshift/oc/pkg/cli/gettoken/pkce"
+)
+
+// Authenticator defines the basic functionality to support Auth Code FLow
+type Authenticator interface {
+	GetTokenByAuthCode(ctx context.Context, input GetTokenByAuthCodeInput, localServerReadyChan chan<- string) (string, string, error)
+	Refresh(ctx context.Context, refreshToken string) (string, string, error)
+	SupportedPKCEMethods() []string
+}
+
+// ExchangeAuthCodeInput uses the fields to exchange auth code and tokens
+type ExchangeAuthCodeInput struct {
+	Code        string
+	PKCEParams  pkce.Params
+	Nonce       string
+	RedirectURI string
+}
+
+// GetTokenByAuthCodeInput stores the fields to get auth code to be used in exchange process.
+type GetTokenByAuthCodeInput struct {
+	BindAddress            []string
+	State                  string
+	Nonce                  string
+	PKCEParams             pkce.Params
+	RedirectURLHostname    string
+	AuthRequestExtraParams map[string]string
+	LocalServerSuccessHTML string
+	LocalServerCertFile    string
+	LocalServerKeyFile     string
+}
+
+type client struct {
+	httpClient           *http.Client
+	provider             *gooidc.Provider
+	oauth2Config         oauth2.Config
+	supportedPKCEMethods []string
+}
+
+func (c *client) GetTokenByAuthCode(ctx context.Context, input GetTokenByAuthCodeInput, localServerReadyChan chan<- string) (string, string, error) {
+	if c.httpClient != nil {
```

Comment:
This is the critical piece of code which nearly all the cred exec plugins have(i.e. [azure](https://github.com/Azure/ARO-RP/blob/aee18fe5709e41757196579edb2cd82d711d1b71/pkg/portal/middleware/aad.go#L282-L284)). Because this is the only way modify the https://cs.opensource.google/go/x/oauth2/+/refs/tags/v0.16.0:oauth2.go;l=343 in oauth2 library to set proxy settings, certificate authorities, timeouts, etc. as in here https://github.com/openshift/oc/pull/1640/files#diff-ed20f9b8315a4eb204230dc820bee5875c99d5cbb6e56234fe28c7ab9d52b87bR149-R161